Calculation of critical angle and Fresnel coefficients

Trave11er opened the pull request:
<!-- BEGIN RELEASE NOTES -->
* physics.optics
   * Added two new functions to `physics.optics.utils`:
   1) `critical_angle() `calculates the onset of total internal reflection, "TIR" when possible (n2 < n1)
   2) `fresnel_coefficients()` calculates real reflection and transmission coefficients when out of TIR and complex ones when the ray is TIR'ed
<!-- END RELEASE NOTES -->
   * Small fix: calculating speed of light in medium is possible given only the refractive index.
   * Associated tests
   * The motivation for the above being that SymPy is well-suited to become a basic 'optics calculator' which would be a go-to tool for small tasks especially given that it can be used via SymPy Live.

Trave11er committed:
Small fix for calculating speed of light in a medium

Trave11er committed:
Implemented calculation of critical angle and Fresnel coefficients

oscargus commented on sympy/physics/optics/tests/test_utils.py:
```diff
@@ -81,11 +93,17 @@ def test_deviation():
 
 def test_brewster_angle():
-    m1 = Medium('m1', permittivity=e0, n=1)
-    m2 = Medium('m2', permittivity=e0, n=1.33)
```
Maybe keep this test as well? That is, just duplicate the assert after these two lines.

oscargus commented on sympy/physics/optics/utils.py:
```diff
@@ -287,6 +353,45 @@ def brewster_angle(medium1, medium2):
+    else:
+        n2 = sympify(medium2)
+
+    if n2 > n1:
```
``if n1 > n2:`? (Based on the error message.)

oscargus commented on sympy/physics/optics/utils.py:
```diff
@@ -171,6 +175,68 @@ def refraction_angle(incident, medium1, medium2, normal=None, plane=None):
 
 
+def fresnel_coefficients(angle_of_incidence, medium1, medium2):
+    # TODO documentation + examples/doctests + tests 
```
This is done it look like?

oscargus commented on sympy/physics/optics/utils.py:
```diff
@@ -171,6 +175,68 @@ def refraction_angle(incident, medium1, medium2, normal=None, plane=None):
+    incidence (labelled 's'). There four real coefficients unless the incident
+    ray reflects in total internal in which case there are two complex ones.
+    Angle of incidence is the angle between the incident ray and the surface 
+    normal. `medium1` and `medium2` can be `Medium` or any sympifiable object.
```
This is far from obvious, but a single back-tick is actually LaTeX in the documentation. You probably want to use two back-ticks.

oscargus commented on sympy/physics/optics/utils.py:
```diff
@@ -171,6 +175,68 @@ def refraction_angle(incident, medium1, medium2, normal=None, plane=None):
+    Parameters
+    ==========
+
+    angle_of_inciddence : sympifiable
```
Spelling of incidence

oscargus commented on sympy/physics/optics/utils.py:
```diff
@@ -171,6 +175,68 @@ def refraction_angle(incident, medium1, medium2, normal=None, plane=None):
+    ==========
+
+    angle_of_inciddence : sympifiable
+    medium1 : sympy.physics.optics.medium.Medium or sympifiable
```
Medium is enough (no need for full path)

oscargus commented on sympy/physics/optics/utils.py:
```diff
@@ -171,6 +175,68 @@ def refraction_angle(incident, medium1, medium2, normal=None, plane=None):
+    angle_of_inciddence : sympifiable
+    medium1 : sympy.physics.optics.medium.Medium or sympifiable
+        Medium 1 or its refractive index
+    medium2 : sympy.physics.optics.medium.Medium or sympifiable
```
Same

oscargus commented:
Thanks!

In general it looks good and is a good addition to Sympy, but a few minor details in the comments to consider and there are some test failures to sort out.

Also, I think it makes sense to remove the three last bullets in the description list. They are clearly valuable information for the PR, but probably shouldn't end up in the release notes. The template provides a good position for these, but right now you can just move them outside of the release note tagged area.

(For the second one I got the impression that it was a change in the test only. In that case the release note can be removed. It makes sense to test both cases though and is a good addition that the default value version is tested.)

Trave11er committed:
Fixed test_utils.py + cosmetic changes

Trave11er commented:
Thank you @oscargus for reviewing the code and detailed feedback
Aapologies for a not-so-clean PR. Have now incorporated all your suggestions and fixed the tests so should be good to go provided the build is fine.

sidhantnagpal commented on sympy/physics/optics/utils.py:
```diff
@@ -171,6 +175,67 @@ def refraction_angle(incident, medium1, medium2, normal=None, plane=None):
+    else:
+        n = n2/n1
+        R_s = simplify((cos(angle_of_incidence)-I*sqrt(sin(angle_of_incidence)**2-n**2))/(cos(angle_of_incidence)+I*sqrt(sin(angle_of_incidence)**2-n**2)))
+        R_p = simplify((n**2*cos(angle_of_incidence)-I*sqrt(sin(angle_of_incidence)**2-n**2))/(n**2*cos(angle_of_incidence)+I*sqrt(sin(angle_of_incidence)**2-n**2)))
```
These lines could possibly be broken down as they are longer than 79 characters. Adding spaces between individual terms separated by operators (like `+`, `-`) will improve the readability of expression.

Is `simplify` necessary for this? It is generally preferable to use only specific simplifications (more information at https://docs.sympy.org/latest/tutorial/simplification.html) in library code, as `simplify` can be slow.

sidhantnagpal commented on sympy/physics/optics/utils.py:
```diff
@@ -171,6 +175,67 @@ def refraction_angle(incident, medium1, medium2, normal=None, plane=None):
+    medium1 : Medium or sympifiable
+        Medium 1 or its refractive index
+    medium2 : Medium or sympifiable
+        Medium 2 or its refractive index
```
The parameters are generally separated by newlines. Like:
```
    angle_of_incidence : sympifiable

    medium1 : Medium or sympifiable
        Medium 1 or its refractive index

    medium2 : Medium or sympifiable
        Medium 2 or its refractive index
```

sidhantnagpal commented on sympy/physics/optics/utils.py:
```diff
@@ -171,6 +175,67 @@ def refraction_angle(incident, medium1, medium2, normal=None, plane=None):
+    incidence (labelled 's'). There four real coefficients unless the incident
+    ray reflects in total internal in which case there are two complex ones.
+    Angle of incidence is the angle between the incident ray and the surface
+    normal. ``medium1`` and ``medium2`` can be ``Medium`` or any sympifiable object.
```
It would be nice to have `References` as well as a `Returns` section in the docstring explaining the conditional output (returned coefficients).

oscargus replied:
Agreed. Wikipedia works well for references.

```
References
==========

.. [1] https:...
```

oscargus replied:
And Returns would come between Parameters and Examples, while References would end the docstring.

sidhantnagpal commented on sympy/physics/optics/__init__.py:
```diff
@@ -30,6 +30,6 @@
 
 from . import utils
-from .utils import (refraction_angle, deviation, brewster_angle, lens_makers_formula,
+from .utils import (refraction_angle, fresnel_coefficients, deviation, brewster_angle, critical_angle, lens_makers_formula,
```
This line could be split as well.

oscargus commented on sympy/physics/optics/utils.py:
```diff
@@ -171,6 +175,67 @@ def refraction_angle(incident, medium1, medium2, normal=None, plane=None):
+    transmission coefficients. Those are obtained for both polarisations
+    when the electric field vector is in the plane of incidence (labelled 'p')
+    and when the electric field vector is perpendicular to the plane of
+    incidence (labelled 's'). There four real coefficients unless the incident
```
"There are four"

Trave11er committed:
Improved documention of fresnel_coefficients()

Trave11er committed:
Changed simplify->cancel in fresnel_coefficients()

Trave11er committed:
Improve readability

Trave11er commented:
Thank you, @oscargus and @sidhantnagpal for further suggestions on documentation and readability - these have been addressed in full.

oscargus commented on sympy/physics/optics/medium.py:
```diff
@@ -120,9 +120,15 @@ def speed(self):
 
         """
-        return 1/sqrt(self._permittivity*self._permeability)
+        if self._permittivity != None and self._permeability != None:
```
Not sure why I didn't think about it earlier, but it is in general better to write `is not None` than `!= None`.

Trave11er replied:
Thank you, done.

Trave11er committed:
Cosmetic

oscargus commented:
Thanks! Just confirming that the tests pass (they should) and then this should be good to merge.

Trave11er commented:
Thanks again @oscargus for the reviewing.

oscargus commented:
Thanks for the contribution!
